src/PaymentGateways/PayPalStandard/Actions/CreatePayPalStandardPaymentURL.php

```diff
@@ -0,0 +1,119 @@
+<?php
+
+namespace Give\PaymentGateways\PayPalStandard\Actions;
+
+use Give\PaymentGateways\DataTransferObjects\GatewayPaymentData;
+use Give\ValueObjects\Address;
+
+/**
+ * This class create PayPal Standard payment gateway one time payment url on basis of donor donation query.
+ *
+ * @unlreased
+ */
+class CreatePayPalStandardPaymentURL
+{
+    public function __invoke(
+        GatewayPaymentData $paymentData,
+        $successRedirectUrl, // PayPal Standard will redirect donor to this url after successful payment.
+        $failedRedirectUrl, // PayPal Standard will redirect donor to this url after failed payment.
+        $payPalIpnListenerUrl // PayPal Standard will send ipn notification to this url on payment update.
+    )
+    {
+        $paypalPaymentRedirectUrl = trailingslashit(give_get_paypal_redirect()) . '?';
+        $itemName = $paymentData->getDonationTitle();
+        $payPalPartnerCode = 'givewp_SP';
+
+        // Setup PayPal API params.
+        $paypalPaymentArguments = [
+            // PayPal account information
+            'business' => give_get_option('paypal_email', false),
+
+            // Donor info
+            'first_name' => $paymentData->donorInfo->firstName,
+            'last_name' => $paymentData->donorInfo->lastName,
+            'email' => $paymentData->donorInfo->email,
+
+            // Donation information.
+            'invoice' => $paymentData->purchaseKey,
+            'amount' => $paymentData->amount,
+            'item_name' => stripslashes($itemName),
+            'currency_code' => give_get_currency($paymentData->donationId),
+
+            // Urls
+            'return' => $successRedirectUrl,
+            'cancel_return' => $failedRedirectUrl,
+            'notify_url' => $payPalIpnListenerUrl,
+
+            'no_shipping' => '1',
+            'shipping' => '0',
+            'no_note' => '1',
+            'charset' => get_bloginfo('charset'),
+            'custom' => $paymentData->donationId,
+            'rm' => '2',
+            'page_style' => give_get_paypal_page_style(),
+            'cbt' => get_bloginfo('name'),
+            'bn' => $payPalPartnerCode,
+        ];
+
+        if ($paymentData->donorInfo->address instanceof Address) {
```

Comment:
I would think that we can assume the type of `donorInfo->address` as provided by the `GatewayPaymentData` class.

Reply:
`donorInfo->address` value can be `false` or `Address` value object. 

I updated the logic to use `billingAddress` which always gives `Address` VO.
fixeD: https://github.com/impress-org/givewp/compare/f4c7843ed9b9...9333cfc25f8e